### PR TITLE
[GOBBLIN-1633] Fix compaction actions on job failure not retried if compaction succeeds

### PR DESCRIPTION
**Pull request by homatthew**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [X] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-1633] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1633

### Description
- [X] Here are some details about my PR, including screenshots (if applicable):
- The issue is that when [CompactionCompleteFileOperationAction.java](https://github.com/apache/gobblin/pull/3494/files#diff-1cff14b642342f6005b830c658cae9c6030148c9c79cf3a914d9acac57f3c644) succeeds , it persists the init time using the key `CompactionSlaEventHelper.LAST_RUN_START_TIME`. However, if the job fails after this point but before metadata registration, then the files will not be registered until the next re-compaction cycle determined by `compaction.timebased.min.recompaction.duration`. 
- Note: This value for the `CompactionSlaEventHelper.LAST_RUN_START_TIME` is used by the [CompactionTimeRangeVerifier.java](https://github.com/apache/gobblin/pull/3494/files#diff-e35a620545631e579301d1d49744d2465729932480c4f9de8eecc0375962b77b) to determine when the last successful compaction started and whether to proceed with recompaction.
- The solution is to persist the value for the key `CompactionSlaEventHelper.LAST_RUN_START_TIME` after the MR compaction task has completed instead of after the file operation is completed. 

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

It is extremely difficult to add tests to the MR Compaction Task due to the tight coupling with the job submission. I have regression tested this using a parallel test pipeline and compacting twice.

The first compaction job successfully compacts the hourly partition to a daily partition, and the second compaction job does not create any work units due to the compaction
```
26-04-2022 11:07:37 PDT daily-compaction-orc INFO - WARN Last compaction for /data/tracking/streaming_parallel_test/PageViewEvent/hourly/2022/04/26 is 2022-04-26T10:51:07.084-07:00, which is not before latestEligibleCompactTime=2022-04-26T10:06:23.429-07:00
```

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-compaction/src/main/java/org/apache/gobblin/compaction/verify/CompactionTimeRangeVerifier.java [hunk 1]
```diff
@@ -99,7 +99,8 @@ public Result verify(FileSystemDataset dataset) {
         if (compactState.contains(CompactionSlaEventHelper.LAST_RUN_START_TIME)
             && compactState.getPropAsLong(CompactionSlaEventHelper.LAST_RUN_START_TIME)
             > latestEligibleCompactTime.getMillis()) {
-          log.warn("Last compaction for {} is {}, not before {}", dataset.datasetRoot(),
+          log.warn("Last compaction for {} is {}, which is not before latestEligibleCompactTime={}",
```
Comment (homatthew) [inline gobblin-compaction/src/main/java/org/apache/gobblin/compaction/verify/CompactionTimeRangeVerifier.java:102]: Readability improvement for log line

File: gobblin-compaction/src/main/java/org/apache/gobblin/compaction/mapreduce/MRCompactionTask.java [hunk 2]
```diff
@@ -101,14 +106,19 @@ public void run() {
...
+        TaskState taskState = taskContext.getTaskState();
+        setCounterInfo(taskState);
 
         List<CompactionCompleteAction> actions = this.suite.getCompactionCompleteActions();
         for (CompactionCompleteAction action: actions) {
           action.addEventSubmitter(eventSubmitter);
           action.onCompactionJobComplete(dataset);
         }
         submitEvent(CompactionSlaEventHelper.COMPACTION_COMPLETED_EVENT_NAME);
+        if (dataset instanceof FileSystemDataset) {
```
Comment (homatthew) [inline gobblin-compaction/src/main/java/org/apache/gobblin/compaction/mapreduce/MRCompactionTask.java:118]: We can only check / persist the state for FileSystemDataset based compaction jobs, so we needs this typecheck here. 

The [CompactionSuiteBaseWithConfigurableCompleteAction.java](https://github.com/apache/gobblin/commit/a77250b057e1ba83531ee0bed6e8302ed6c18510#diff-3befde2a879553d0eb7deb8a0c9b48951dc1c26c6bcf9fe86701bbfaf3c295df) and [CompactionSuiteBase.java](https://github.com/apache/gobblin/blob/d9ae5353c74fdcd385835fca9b586b3fdb90971b/gobblin-compaction/src/main/java/org/apache/gobblin/compaction/suite/CompactionSuiteBase.java) both implement `CompactionSuite<FileSystemDataset>`.

File: gobblin-compaction/src/main/java/org/apache/gobblin/compaction/action/CompactionCompleteFileOperationAction.java [hunk 3]
```diff
@@ -183,8 +182,6 @@ public void onCompactionJobComplete(FileSystemDataset dataset) throws IOExceptio
           this.configurator.getConfiguredJob().getJobID().toString());
       compactionState.setProp(DUPLICATE_COUNT_TOTAL,
           job.getCounters().findCounter(RecordKeyDedupReducerBase.EVENT_COUNTER.DEDUPED).getValue());
-      compactionState.setProp(CompactionSlaEventHelper.LAST_RUN_START_TIME,
-          this.state.getProp(CompactionSource.COMPACTION_INIT_TIME));
```
Comment (homatthew) [inline gobblin-compaction/src/main/java/org/apache/gobblin/compaction/action/CompactionCompleteFileOperationAction.java:187]: Code moved to the MrCompactionTask

**Comment by vikrambohra** [comment]
LGTM and thank you for putting the effort to fix the typos!

**Review by ZihanLi58 — APPROVED** [review]
+1, thanks for the fix

**Pull request merged**